tests/subsys/net/lib/nrf_cloud/cloud/src/main.c

```diff
@@ -0,0 +1,1880 @@
+/* Verify that nrf_cloud_disconnect function will
+ * fail when the cloud transport fail to disconnect.
+ */
+ZTEST(nrf_cloud_test, test_disconnect_transport_failed)
+{
+	zassert_equal(STATE_IDLE, nfsm_get_current_state(),
+		"nrf_cloud lib should be in the idle state (uninitialized) at the start of test");
+
+	/* Init the cloud successfully */
+	init_cloud_success();
+
+	/* Connect to the cloud */
+	nct_connect_fake.custom_fake = fake_nct_connect__succeeds;
+	/* Transport failed when disconnecting */
+	nct_disconnect_fake.custom_fake = fake_nct_disconnect__fails;
+	/* These fakes are for the poll thread */
+	nct_socket_get_fake.custom_fake = fake_nct_socket_get_data;
+	nct_keepalive_time_left_fake.custom_fake = fake_nct_keepalive_time_left_get;
+	nct_process_fake.custom_fake = fake_nct_process__succeeds;
+	(void)nrf_cloud_connect();
+
+	/* Wait for the connected event for 2s, timeout after that */
+	wait_counter_t = 10;
+	while (--wait_counter_t && !atomic_get(&cloud_connected_t)) {
+		k_sleep(K_MSEC(200));
+	}
+	zassert_not_equal(0, wait_counter_t, "cloud connect timed out");
+
+	zassert_equal(STATE_CONNECTED, nfsm_get_current_state(),
+		"nrf_cloud lib should be in the connected state after successfully connected");
+
+	int ret = nrf_cloud_disconnect();
+
+	zassert_not_equal(0, ret,
+		"return should not be 0 when disconnecting with the transport failed");
+}
+
+/* Verify that nrf_cloud_disconnect function will succeed
+ * given the cloud transport disconnected without errors.
+ */
+ZTEST(nrf_cloud_test, test_disconnect_success)
+{
+	zassert_equal(STATE_IDLE, nfsm_get_current_state(),
+		"nrf_cloud lib should be in the idle state (uninitialized) at the start of test");
+
+	/* Init the cloud successfully */
+	init_cloud_success();
+
+	/* Connect to the cloud successfully */
+	connect_cloud_success();
+
+	int ret = nrf_cloud_disconnect();
+
+	/* Wait for the disconnected event for 2s, timeout after that */
+	wait_counter_t = 10;
+	while (--wait_counter_t && atomic_get(&cloud_connected_t)) {
+		k_sleep(K_MSEC(200));
+	}
+	zassert_not_equal(0, wait_counter_t, "cloud disconnect timed out");
+
+	zassert_ok(ret,	"return should be 0 when success");
+	zassert_equal(STATE_INITIALIZED, nfsm_get_current_state(),
+		"nrf_cloud lib should be in the initialized state after disconnected successfully");
+}
+
+/* TEST NRF_CLOUD_SHADOW_DEVICE_STATUS_UPDATE */
+
+/* Verify that the function will fail if the
+ * internal state of the lib is not dc_connected
+ */
+ZTEST(nrf_cloud_test, test_shadow_device_status_update_not_dc_connected)
+{
+	/* Data for shadow device status update */
+	struct nrf_cloud_device_status dev_status_t = {
+		.modem = NULL,
+		.svc = NULL
+	};
+
+	/* Custom fakes for nrf_cloud_send */
+	nct_cc_send_fake.custom_fake = fake_nct_cc_send__succeeds;
+	nct_dc_stream_fake.custom_fake = fake_nct_dc_stream__succeeds;
+	nct_dc_send_fake.custom_fake = fake_nct_dc_send__succeeds;
+	nct_dc_bulk_send_fake.custom_fake = fake_nct_dc_bulk_send__succeeds;
```

Comment:
The minute we start involving nct_cc_send and nct_dc_send / nrf_cloud_send is the minute I think mocking MQTT with fake messages from backend itself becomes more appropriate

As I've written in other comments, in my opinion, these shadow tests should be in a new test suite that is based on MQTT mocking, and probably part of a different PR/ticket altogether